rule_info/rule_info.proto

```diff
@@ -0,0 +1,77 @@
+// Proto messages that Haskell build rules output, for use by tools like
+// "hrepl".
+syntax = "proto3";
+
+package haskell;
+
+// Information about a HaskellCompile action.
+message CompileInfo {
+  reserved 1, 3, 4, 7, 8, 10, 13, 14;
+  // The source files that are being compiled (.hs and .lhs files).
+  // Does not contain "boot" files (e.g. ".hs-boot").
+  repeated string source_files = 2;
+  // Modules that should not be exposed.  (For libraries only.)
+  repeated string hidden_modules = 12;
+  // The runfiles that should be collected for this target.
+  repeated Runfile runfiles = 5;
+  // The Haskell package name.  Should only be set for third-party libraries
+  // (namely, those with the "cabal_version" attribute).
+  string package_name = 6;
+
+  // Arguments to GHC for compiling this target.  Does not include source files,
+  // nor flags related to dependent packages ("-package-id" and "-package-db").
+  repeated string options = 9;
+
+  // Shared libraries for cc_library rules that this target depends on
+  // transitively.
+  repeated string transitive_cc_shared_libs = 11;
+}
+
+// Information about a built library.
+message LibraryInfo {
```

Comment:
It doesn't currently handle them.  We originally used static archives, but switched a year ago to dynamically-linked GHC to work around bugs in the static loader.   It seems pretty reasonable to me that it could support both.

I filed google/hrepl#7 to track this, though as you've noted it would also require changes to rules_haskell itself.

Reply:
@judah do you remember what bugs those were? We're likewise interested in making the static linker work better for everyone. It would be quite a win: in the common case we could avoid building dynamic libraries altogether.

Reply:
> though as you've noted it would also require changes to rules_haskell itself.

@judah which required changes are you referring to?

`rules_haskell` supports static rts GHC and [sets `-fPIC` and `-fexternal-dynamic-refs` as required](https://github.com/tweag/rules_haskell/blob/d76a16eb2a21b184bfda8dbbc05ee4c1c2239015/haskell/private/actions/compile.bzl#L270) and uses `pic_static_library` of dependencies where available. On Windows that's the default, as the Windows bindist uses static rts mode. On Unix it's optional, currently exposed for nixpkgs provisioned GHC through the [`is_static` attribute](https://api.haskell.build/haskell/nixpkgs.html#haskell_register_ghc_nixpkgs), since there are no static rts GHC bindists available for Unix.

For example, Digital Asset's open source [daml repository](https://github.com/digital-asset/daml/blob/5231bab42399b515c3ae64d8836e2dde3c6a33c7/WORKSPACE#L323) uses static linking in this way. As a test, I've run `hrepl` with this PR's patches on that repository and was able to load a large target with static C library dependencies into GHCi using
```
LIBRARY_PATH=$PWD/bazel-out/k8-opt/bin/compiler/damlc/_ghci_libs-damlc hrepl damlc
```

Looking at `cc_libs` in `compile_info_output_groups`, it doesn't seem to distinguish static and dynamic libraries, so `transitive_cc_shared_libs` will end up holding both. So, maybe just the name is inaccurate.

Reply:
Ah, I didn't realize that static libraries got passed through the same field of cc_libs.  I've renamed the field to `transitive_cc_libs`.  (I don't have access to a static GHC at the moment, so we may need to leave that to future work.)

We got hit by a few bugs around the static linker in the transition from 8.0->8.4 and 8.4->8.6, including multiple instances of segfaults on ANN pragmas.  One example was:
https://gitlab.haskell.org/ghc/ghc/issues/14675
But there were a couple others we never tracked down well enough to file a useful bug report (in particular we weren't able to rule out effects of our internal toolchain).

At the time (early 2019), the existing comments on the GHC issues didn't give us a lot of confidence in the static RTS. But it's good to hear you've taken it up in earnest, and your use of `-fexternal-dynamic-refs` sounds promising; we never discovered that flag, and the problem mentioned in that link (`R_X86_64_PC32`) does sound familiar. I'll make a note to try using that in our codebase.

Reply:
FWIW my understanding is that interest in the static linker from GHC HQ has varied over the years. When support from dynamic libraries was first introduced, I suppose the vision was to eventually retire the static linker, because it's duplicating so much OS functionality. However, the story around dynamic libraries turned out less than ideal. And at this point, I believe even Facebook (who employs @simonmar) might be using a static toolchain, or so https://engineering.fb.com/security/fighting-spam-with-haskell/ would have me believe.

At any rate, the static toolchain is what Digital Asset uses too.

Reply:
Actually I believe we're not using the static runtime linker any more at Facebook. The static runtime linker requires all linked objects to fit in the low 2GB of the address space, and we eventually ran into that limit and had to switch to building a standard `.so` and using the system dyanmic linker to load it.

Reply:
@simonmar very interesting. And thanks for chiming in! So out of curiosity, do you now _only_ use standard .so during your build, or do you still use `-dynamic-too` when building? If the former, that would be nice to speed up the build, but it's not an option for our macOS users, where the Mach-O header size limits are intentionally very low (hence, there's only so many dynamic dependencies one can have).